File: samples/ocr_e2e/ocr_e2e_async.py
#
# Copyright (C) 2025 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#
import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)


import cv2
import argparse
from ocr_e2e_async_cls import OCR_e2e_Async

import vaststreamx as vsx
import threading
from easydict import EasyDict as edict

attr = vsx.AttrKey
import time
import ast


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--det_model",
        default="/opt/vastai/vaststreamx/data/models/dbnet_resnet50_vd-int8-kl_divergence-1_3_736_1280-vacc/mod",
        help="text detection model prefix of the model suite files",
    )
    parser.add_argument(
        "--det_vdsp_params",
        default="./data/configs/dbnet_rgbplanar.json",
        help="text detection vdsp preprocess parameter file",
    )
    parser.add_argument(
        "--det_box_type",
        default="quad",
        help="det box type, poly or quad",
    )
    parser.add_argument(
        "--det_elf_file",
        default="/opt/vastai/vaststreamx/data/elf/find_contours_ext_op",
        help="input file",
    )
    parser.add_argument(
        "--cls_model",
        default="/opt/vastai/vaststreamx/data/models/resnet34_vd-int8-max-1_3_32_100-vacc/mod",
        help="text detection model prefix of the model suite files",
    )
    parser.add_argument(
        "--cls_vdsp_params",
        default="./data/configs/crnn_rgbplanar.json",
        help="text detection vdsp preprocess parameter file",
    )
    parser.add_argument(
        "--cls_label_list",
        type=list,
        default=["0", "180"],
        help="text classification label list",
    )
    parser.add_argument(
        "--cls_thresh", type=float, default=0.9, help="text classification thresh"
    )
    parser.add_argument(
        "--rec_model",
        default="/opt/vastai/vaststreamx/data/models/resnet34_vd-int8-max-1_3_32_100-vacc/mod",
        help="text detection model prefix of the model suite files",
    )
    parser.add_argument(
        "--rec_vdsp_params",
        default="./data/configs/crnn_rgbplanar.json",
        help="text detection vdsp preprocess parameter file",
    )
    parser.add_argument(
        "--rec_label_file",
        default="../data/labels/key_37.txt",
        help="text recognizition label file",
    )
    parser.add_argument(
        "--rec_drop_score",
        type=float,
        default=0.5,
        help="text recogniztion drop score threshold",
    )
    parser.add_argument(
        "--use_angle_cls", type=bool, default=True, help="whether use angle classifier"
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--device_ids",
        default="[0]",
        type=str,
        help="device ids to run",
    )
    parser.add_argument(
        "--input_file",
        default="../../../data/images/detect.jpg",
        help="input file",
    )
    parser.add_argument(
        "--output_file",
        default="./ocr_result.jpg",
        help="output file",
    )
    parser.add_argument(
        "--dataset_filelist",
        default="",
        help="dataset filelist",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    parser.add_argument(
        "--dataset_output_file",
        default="",
        help="dataset output file",
    )
    args = parser.parse_args()
    return args


def inference_async(ocr_e2e, args, context, thread_index):
    vsx.set_device(ocr_e2e.device_id)

    ticks = []
    tocks = []

    # get output thread
    def get_output_thread(ocr_e2e, filelist):
        vsx.set_device(ocr_e2e.device_id)
        index = 0
        while True:
            try:
                output = ocr_e2e.get_output()
                tocks.append(time.time())
                print(f"Thread:{thread_index},Get {filelist[index]} result")
                if len(filelist) == 1:
                    boxes, rec_res = output
                    for box, rec_result in zip(boxes, rec_res):
                        str = "["
                        for point in box:
                            str += f"[{point[0]},{point[1]}], "
                        str = str[:-2] + "], "
                        print(str, rec_result)
                    if args.output_file != "":
                        cv_mat = cv2.imread(filelist[index])
                        for box in boxes:
                            for i in range(len(box)):
                                t = (i + 1) % len(box)
                                pt1 = (box[i][0], box[i][1])
                                pt2 = (box[t][0], box[t][1])
                                cv2.line(cv_image, pt1, pt2, color=(0, 0, 255))
                        direc, filename = os.path.split(args.output_file)
                        outfile = os.path.join(
                            direc, f"thread_{thread_index}_{filename}"
                        )
                        cv2.imwrite(outfile, cv_mat)
                        print(f"save file to {outfile}")
                index += 1
            except Exception:
                break

    # one image test
    if args.dataset_filelist == "":
        cv_mat = cv2.imread(args.input_file)
        assert cv_mat is not None, f"Failed to read image file: {args.input_file}"
        output_thread = threading.Thread(
            target=get_output_thread, args=(ocr_e2e, [args.input_file])
        )
        output_thread.start()

        ocr_e2e.process_async(cv_mat)

        ocr_e2e.stop()
        output_thread.join()
        return

    # dataset test
    filelist = []
    with open(args.dataset_filelist, "rt") as f:
        files = f.readlines()
    if args.dataset_root != "":
        for file in files:
            file = file.replace("\n", "")
            filelist.append(os.path.join(args.dataset_root, file))
    else:
        filelist = files

    output_thread = threading.Thread(target=get_output_thread, args=(ocr_e2e, filelist))
    output_thread.start()

    for file in filelist:
        cv_image = cv2.imread(file)
        assert cv_image is not None, f"Failed to read image file:{file}"
        ocr_e2e.process_async(cv_image)
        ticks.append(time.time())

    ocr_e2e.stop()
    output_thread.join()

    context.merge_lock.acquire()
    context.ticks += ticks
    context.tocks += tocks
    context.merge_lock.release()


if __name__ == "__main__":
    args = argument_parser()
    device_ids = ast.literal_eval(args.device_ids)

    models = []
    for id in device_ids:
        ocr_e2e = OCR_e2e_Async(
            args.det_model,
            args.det_vdsp_params,
            args.det_box_type,
            args.det_elf_file,
            args.cls_model,
            args.cls_vdsp_params,
            args.cls_label_list,
            args.cls_thresh,
            args.rec_model,
            args.rec_vdsp_params,
            args.rec_label_file,
            args.rec_drop_score,
            args.use_angle_cls,
            batch_size=1,
            device_id=id,
            hw_config=args.hw_config,
        )
        models.append(ocr_e2e)

    threads = []
    context = edict(
        merge_lock=threading.Lock(),
        ticks=[],
        tocks=[],
    )

    start = time.time()
    for mod, id in zip(models, device_ids):
        thread = threading.Thread(target=inference_async, args=(mod, args, context, id))
        thread.start()
        threads.append(thread)

    for thread in threads:
        thread.join()
    test_cost = time.time() - start

    if args.dataset_filelist == "":
        exit(0)

    ticks = context.ticks
    tocks = context.tocks

    cost_sum = 0
    assert len(ticks) == len(
        tocks
    ), f"ticks len = {len(ticks)}, tocks len = {len(tocks)}"
    for i in range(len(ticks)):
        cost_sum += tocks[i] - ticks[i]

    avg_cost = cost_sum / len(ticks)

    throughput = len(ticks) / test_cost

    print(
        f"Image count: {len(ticks)}, total cost: {test_cost:.2f} s, throughput: {throughput:.2f} fps, average latency: {avg_cost:.3f} s"
    )

File: samples/ocr_e2e/ocr_e2e_async.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <chrono>
#include <mutex>

#include "common/cmdline.hpp"
#include "common/file_system.hpp"
#include "common/utils.hpp"
// #include "ocr_e2e_multi_thread.hpp"
#include "ocr_e2e_async.hpp"
#include "opencv2/opencv.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>(
      "det_model", '\0', "text detection model prefix of the model suite files",
      false, "/opt/vastai/vaststreamx/data/models/det_model_vacc_fp16/mod");
  args.add<std::string>("det_config", '\0',
                        "text detection vdsp preprocess parameter file", false,
                        "../data/configs/dbnet_rgbplanar.json");
  args.add<std::string>(
      "cls_model", '\0',
      "text classification model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/cls_model_vacc_fp16/mod");
  args.add<std::string>("cls_config", '\0',
                        "text classification vdsp preprocess parameter file",
                        false, "../data/configs/crnn_rgbplanar.json");
  args.add<std::string>(
      "rec_model", '\0',
      "text recognition model prefix of the model suite files", false,
      "/opt/vastai/vaststreamx/data/models/rec_model_vacc_fp16/mod");
  args.add<std::string>("rec_config", '\0',
                        "text recognition vdsp preprocess parameter file",
                        false, "../data/configs/crnn_rgbplanar.json");

  args.add<std::string>("det_box_type", '\0', "text detection box type", false,
                        "quad");
  args.add<std::string>(
      "det_elf_file", '\0', "text detection elf file", false,
      "/opt/vastai/vaststreamx/data/elf/find_contours_ext_op");
  args.add<std::string>("cls_labels", '\0', "text classification label list",
                        false, "[0, 180]");
  args.add<float>("cls_thresh", '\0', "text classification thresh", false, 0.9);
  args.add<std::string>("rec_label_file", '\0', "text recognition label file",
                        false, "../data/labels/ppocr_keys_v1.txt");
  args.add<float>("rec_drop_score", '\0',
                  "text recogniztion drop score threshold", false, 0.5);
  args.add<int>("use_angle_cls", '\0', "use text classification", false, 1);
  args.add<uint32_t>("batch_size", '\0', "batch size of the model", false, 1);
  args.add<std::string>("device_ids", '\0', "device id to run", false, "[0]");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("input_file", '\0', "input image", false,
                        "../data/images/word_336.png");
  args.add<std::string>("output_file", '\0', "output image file", false, "");
  args.add<std::string>("dataset_filelist", '\0', "input dataset filelist",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.add<std::string>("dataset_output_file", '\0', "dataset output file",
                        false, "");
  args.add<uint32_t>("queue_size", '\0', "set queue size", false, 1);

  args.parse_check(argc, argv);
  return args;
}

void InferenceThread(std::shared_ptr<vsx::OCR_e2e_Async> model,
                     cmdline::parser& args, std::mutex& merge_mutex,
                     std::vector<int64_t>& costs, uint32_t device_id) {
  std::vector<time_point> ticks;
  std::vector<time_point> tocks;
  std::vector<std::string> filelist;
  vsx::SetDevice(device_id);
  // get output
  auto output_thread = std::thread([&]() {
    vsx::SetDevice(device_id);
    int index = 0;
    while (true) {
      std::vector<vsx::TextObject> text_objs;
      if (model->GetOutput(text_objs)) {
        tocks.push_back(std::chrono::high_resolution_clock::now());
        std::cout << "Thread " << device_id << " get " << filelist[index]
                  << " result.\n";
        if (filelist.size() == 1) {
          for (auto& item : text_objs) {
            auto coor = std::get<0>(item);
            auto score = std::get<1>(item);
            auto str = std::get<2>(item);
            std::cout << "bbox:[ [" << static_cast<int>(coor[0]) << " "
                      << static_cast<int>(coor[1]) << "] ["
                      << static_cast<int>(coor[2]) << " "
                      << static_cast<int>(coor[3]) << "] ["
                      << static_cast<int>(coor[4]) << " "
                      << static_cast<int>(coor[5]) << "] [ "
                      << static_cast<int>(coor[6]) << " "
                      << static_cast<int>(coor[7]) << "] ], score: " << score
                      << ", string: " << str << std::endl;
          }
          // draw box to image
          if (!args.get<std::string>("output_file").empty()) {
            auto cv_image = cv::imread(filelist[0]);
            for (auto& item : text_objs) {
              auto coor = std::get<0>(item);
              auto str = std::get<2>(item);
              cv::line(cv_image, cv::Point2f(coor[0], coor[1]),
                       cv::Point2f(coor[2], coor[3]), cv::Scalar(0, 0, 255));
              cv::line(cv_image, cv::Point2f(coor[2], coor[3]),
                       cv::Point2f(coor[4], coor[5]), cv::Scalar(0, 0, 255));
              cv::line(cv_image, cv::Point2f(coor[4], coor[5]),
                       cv::Point2f(coor[6], coor[7]), cv::Scalar(0, 0, 255));
              cv::line(cv_image, cv::Point2f(coor[0], coor[1]),
                       cv::Point2f(coor[6], coor[7]), cv::Scalar(0, 0, 255));
            }
            fs::path output_path = args.get<std::string>("output_file");
            auto dir = output_path.parent_path().string();
            if (dir.empty()) dir = ".";
            auto filename = output_path.filename().string();
            auto save_file =
                dir + "/thread_" + std::to_string(device_id) + "_" + filename;
            std::cout << "Save file to: " << save_file << std::endl;
            cv::imwrite(save_file, cv_image);
          }
        }
        index++;
      } else {
        break;
      }
    }
  });

  // one image test
  if (args.get<std::string>("dataset_filelist").empty()) {
    filelist.push_back(args.get<std::string>("input_file"));
    auto cv_image = cv::imread(filelist[0]);
    model->ProcessAsync(cv_image);
    model->Stop();
    output_thread.join();
    return;
  }

  // dataset test
  std::vector<std::string> namelist =
      vsx::ReadFileList(args.get<std::string>("dataset_filelist"));
  auto dataset_root = args.get<std::string>("dataset_root");
  for (size_t s = 0; s < namelist.size(); s++) {
    auto fullname = namelist[s];
    if (!dataset_root.empty()) fullname = dataset_root + "/" + fullname;
    filelist.push_back(fullname);
  }

  // set input
  for (size_t s = 0; s < filelist.size(); s++) {
    auto cv_image = cv::imread(filelist[s]);
    model->ProcessAsync(cv_image);
    ticks.push_back(std::chrono::high_resolution_clock::now());
  }
  // set stop
  model->Stop();
  // wait thread done
  output_thread.join();

  if (ticks.size() != tocks.size()) {
    std::cout << "Error! ticks.size() != tocks.szie(). ticks.size() = "
              << ticks.size() << ", tocks.size() = " << tocks.size()
              << std::endl;
  }

  merge_mutex.lock();
  for (size_t i = 0; i < ticks.size(); i++) {
    auto cost = std::chrono::duration_cast<std::chrono::milliseconds>(tocks[i] -
                                                                      ticks[i])
                    .count();
    costs.push_back(cost);
  }
  merge_mutex.unlock();
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  auto device_ids = vsx::ParseVecUint(args.get<std::string>("device_ids"));
  auto cls_labels = vsx::ParseVecUint(args.get<std::string>("cls_labels"));

  std::vector<std::shared_ptr<vsx::OCR_e2e_Async>> models;
  models.reserve(device_ids.size());
  for (auto device_id : device_ids) {
    auto model = std::make_shared<vsx::OCR_e2e_Async>(
        args.get<std::string>("det_model"), args.get<std::string>("det_config"),
        args.get<std::string>("det_box_type"),
        args.get<std::string>("det_elf_file"),
        args.get<std::string>("cls_model"), args.get<std::string>("cls_config"),
        cls_labels, args.get<float>("cls_thresh"),
        args.get<std::string>("rec_model"), args.get<std::string>("rec_config"),
        args.get<std::string>("rec_label_file"),
        args.get<float>("rec_drop_score"), args.get<int>("use_angle_cls"),
        args.get<uint32_t>("batch_size"), device_id,
        args.get<std::string>("hw_config"), args.get<uint32_t>("queue_size"));
    models.push_back(model);
  }

  std::mutex merge_mutex;
  std::vector<int64_t> costs;
  std::vector<std::thread> threads;
  threads.reserve(device_ids.size());

  auto start = std::chrono::high_resolution_clock::now();
  for (size_t i = 0; i < device_ids.size(); i++) {
    threads.emplace_back(InferenceThread, models[i], std::ref(args),
                         std::ref(merge_mutex), std::ref(costs), device_ids[i]);
  }

  for (auto& thread : threads) {
    thread.join();
  }
  auto end = std::chrono::high_resolution_clock::now();
  auto total_cost =
      std::chrono::duration_cast<std::chrono::milliseconds>(end - start)
          .count();

  // test one image
  if (args.get<std::string>("dataset_filelist").empty()) {
    return 0;
  }

  // test dataset
  int64_t cost_sum = 0;
  for (auto cost : costs) cost_sum += cost;

  std::cout << "Image count: " << costs.size() << ", total cost: " << total_cost
            << " ms, throughput: " << costs.size() * 1000.0 / total_cost
            << " fps. Average latency: " << cost_sum * 1.0 / costs.size()
            << " ms. " << std::endl;

  return 0;
}